aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -359,6 +361,14 @@ CUDAGraph::~CUDAGraph() {
     generator_state->unregister_graph(this);
   }
   reset();
+
+// There are recent HIP changes where hipGraphExecDestroy doesn't immediately free memory.
+//  They wait for next sync point in order to free the memory, this is to ensure that all
+//  hipGraphLaunch are finished before we release any memory. This feature was enabled in rocm6.2.  
+#if (defined(ROCM_VERSION) && ROCM_VERSION >= 60200)
```

Comment:
Is the extra sync required now?

Reply:
Yes, as cudaGraphInstantiateFlagAutoFreeOnLaunch only adds async frees after each launch, we need to ensure all async opreations finish before destroying the graph. 

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cudagraphinstantiateflagautofreeonlaunch

Reply:
I'm not sure I understand why this is a ROCm-only logic?

Reply:
@jeffdaily @pruthvistony I had an offline discussion with @pragupta on this, and I feel like it's okay to merge this patch into our ROCm fork as being a ROCm-conditional one, but for the upstream PR, I'd rather see it unconditional since our understanding is that this issue affects CUDA as well, if the flow reaches the destructor before the async frees are executed. For the upstream PR, I suggested Prachi to create a new test case that generates this scenario, so we can easily justify why these extra syncs are needed. Let me know your thoughts.